### DISCUSSION TITLE
Merge Path + Query Parameters in BaseModel

### DISCUSSION THREAD

**Discussion by CrashLaker** [fastapi/fastapi#7311] [Questions] [answered]
Hi all,

I just started using fastapi and I'm struggling to understant how to use the BaseModel on both path + query parameters.

my endpoint:
```
rs = client.get("/data/a/b/c")
# possibly
rs = client.get("/data/a/b/c?sort=true&page=1")
```
On the code below is there a way to make the item return all the optional values declared in the Model?
```python
class Item(BaseModel):
    sort: Optional[str] =  ""
    page: Optional[int] = 1
    per_page: Optional[int] = 20
    search_term: Optional[str] = ""

@app.get("/data/{l1}/{l2}/{l3}")
def d_tree(
           l1: str,
           l2: str,
           l3: str,
           item: Optional[Item] = None):
    print(l1, l2, l3) # l1 l2 l3 ok
    print(item) # None
```

This would throw:
```python
@app.get("/data/{l1}/{l2}/{l3}")
def d_tree(*,
           l1: str,
           l2: str,
           l3: str,
           item: Item):
#{'detail': [{'loc': ['body'], 'msg': 'field required', 'type': 'value_error.missing'}]}
```

I'm wondering if It's possible to get all my variables inside the Model like this:
```python
class Item(BaseModel):
    l1: str
    l2: str
    l3: str
    sort: Optional[str] =  ""
    page: Optional[int] = 1
    per_page: Optional[int] = 20
    search_term: Optional[str] = ""

@app.get("/data/{l1}/{l2}/{l3}")
def d_tree(item: Item):
    toil.phead(item)

    return []
```

### Result
```
{'detail': [{'loc': ['body'], 'msg': 'field required', 'type': 'value_error.missing'}]}
```

### Expect
```
{   
    'l1': 'a',
    'l2': 'b',
    'l3': 'c',
    'page': 1,
    'per_page': 20,
    'search_term': '',
    'sort': ''
}
```

**Comment by jrversteegh** [comment]
Possibly, but it doesn't look like a particularly good design. Query and path parameters typically serve a different purpose. Having both in one model at least violates the Single Responsibility principle.

**Comment by CrashLaker** [comment]
Hi @jrversteegh,

thank you for your reply.
I understood that indeed isn't a good design.
That said, is it possible to at least return the item object value here with all the default options set?
```python
class Item(BaseModel):
    sort: Optional[str] =  ""
    page: Optional[int] = 1
    per_page: Optional[int] = 20
    search_term: Optional[str] = ""

@app.get("/data/{l1}/{l2}/{l3}")
def d_tree(
           l1: str,
           l2: str,
           l3: str,
           item: Optional[Item] = None):
    print(l1, l2, l3) # l1 l2 l3 ok
    print(item) # None  <---------------- here
```
I was expecting item to be:
```python
item = {
  sort: '',
  page: 1,
  per_page: 20,
  search_term: '',
}
```

**Comment by jrversteegh** [comment, marked as answer]
I think that doesn't work because Query parameters need to be scalars or a list of scalars. They can't (yet) be objects, but for a flat object like yours, you could do something like:

```python
@dataclass
class Item:
    sort: str =  ""
    page: int = 1
    per_page: int = 20
    search_term: str = ""

@app.get("/data/{l1}/{l2}/{l3}")
def d_tree(
           l1: str,
           l2: str,
           l3: str,
           item: Item = Depends()):
    print(l1, l2, l3) # l1 l2 l3 ok
    print(item) # Should work now
```
Note the `dataclass` decorator, which makes sure that `Item` has an initializer for all its fields.

**Comment by CrashLaker** [comment]
Hi @jrversteegh,

thank you! it worked :)
Although I'm trying to understand why sort ended up as "True" whereas search_term had the correct value empty string.
```
Item(sort='true', page=1, per_page=20, search_term='')
```

**Comment by jrversteegh** [comment]
@CrashLaker because you requested `/data/ham/eggs/spam?sort=true` ?

**Comment by CrashLaker** [comment]
@jrversteegh  err...  yeah nvm
that was dumb.. i'm so sorry. all sorted out now thank you so much :)

**Comment by jrversteegh** [comment]
np and you're welcome.

**Comment by tiangolo** [comment]
Thanks for the help here @jrversteegh! 🚀 

And thanks @CrashLaker for coming back to close the issue.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
